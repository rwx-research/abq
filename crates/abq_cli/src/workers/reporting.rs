use abq_reporting::{CompletedSummary, Reporter, ReportingError};
use abq_utils::{
    log_assert,
    net_protocol::{client::ReportedResult, queue::AssociatedTestResults},
    results_handler::{NotifyResults, SharedNotifyResults, SharedResultsHandler},
};
use async_trait::async_trait;
use tokio::{sync::mpsc, task::JoinHandle};

use crate::reporting::{reporter_from_kind, ReporterKind, StdoutPreferences, SuiteResult};

use super::summary;

struct ReportingTask {
    rx_results: mpsc::UnboundedReceiver<Vec<AssociatedTestResults>>,
    reporters: Vec<Box<dyn Reporter>>,
    overall_tracker: summary::SuiteTracker,
}

pub(crate) struct FinalizedReporters {
    suite_result: SuiteResult,
    reporters: Vec<Box<dyn Reporter>>,
}

impl ReportingTask {
    fn new(
        rx_results: mpsc::UnboundedReceiver<Vec<AssociatedTestResults>>,
        reporters: Vec<Box<dyn Reporter>>,
    ) -> Self {
        Self {
            reporters,
            rx_results,
            overall_tracker: Default::default(),
        }
    }

    async fn work_forever(mut self) -> FinalizedReporters {
        while let Some(results) = self.rx_results.recv().await {
            results
                .into_iter()
                .for_each(|results| self.handle_associated_results(results));
        }

        self.reporters
            .iter_mut()
            .for_each(|reporter| reporter.after_all_results());

        FinalizedReporters {
            suite_result: self.overall_tracker.suite_result(),
            reporters: self.reporters,
        }
    }

    fn handle_associated_results(&mut self, results: AssociatedTestResults) {
        let AssociatedTestResults {
            work_id: _,
            run_number,
            results,
            before_any_test,
            after_all_tests,
        } = results;

        assert!(
            !results.is_empty(),
            "ABQ protocol must never ship empty list of test results"
        );

        let mut output_before = Some(before_any_test);
        let mut output_after = after_all_tests;

        let mut results = results.into_iter().peekable();

        while let Some(test_result) = results.next() {
            self.overall_tracker
                .account_result(run_number, &test_result);

            let output_before = std::mem::take(&mut output_before);
            let output_after = if results.peek().is_none() {
                // This is the last test result, the output-after is associated
                // with it
                std::mem::take(&mut output_after)
            } else {
                None
            };

            let reported_result = ReportedResult {
                output_before,
                test_result,
                output_after,
            };

            for reporter in self.reporters.iter_mut() {
                // TODO: is there a reasonable way to surface the error?
                let _opt_error = reporter.push_result(run_number, &reported_result);
            }
        }
    }
}

impl FinalizedReporters {
    pub fn finish(self, summary: &CompletedSummary) -> (SuiteResult, Vec<ReportingError>) {
        let Self {
            suite_result,
            reporters,
        } = self;

        let errors: Vec<_> = reporters
            .into_iter()
            .filter_map(|reporter| reporter.finish(summary).err())
            .collect();

        (suite_result, errors)
    }
}

pub(crate) struct ReportingTaskHandle {
    handle: JoinHandle<FinalizedReporters>,
}

impl ReportingTaskHandle {
    pub async fn join(self) -> FinalizedReporters {
        self.handle
            .await
            .expect("reporting task must not be dropped before its handle is")
    }
}

#[derive(Clone)]
pub struct ReportingProxy {
    tx_results: mpsc::UnboundedSender<Vec<AssociatedTestResults>>,
}

#[async_trait]
impl NotifyResults for ReportingProxy {
    async fn send_results(&mut self, results: Vec<AssociatedTestResults>) {
        let res = self.tx_results.send(results);
        log_assert!(
            res.is_ok(),
            "reporting receiver died before all test results consumed"
        );
    }
}

impl SharedNotifyResults for ReportingProxy {
    fn boxed_clone(&self) -> SharedResultsHandler {
        Box::new(self.clone())
    }
}

pub fn build_reporters(
    reporter_kinds: Vec<ReporterKind>,
    stdout_preferences: StdoutPreferences,
    test_suite_name: &str,
) -> Vec<Box<dyn Reporter>> {
    reporter_kinds
        .into_iter()
        .map(|kind| reporter_from_kind(kind, stdout_preferences, test_suite_name))
        .collect()
}

/// Spawns a task to handle reporting of results.
/// Returns a [ReportingProxy] that can be used to communicate with the reporting task.
pub(crate) fn create_reporting_task(
    reporters: Vec<Box<dyn Reporter>>,
) -> (ReportingProxy, ReportingTaskHandle) {
    // TODO make the channel bounded, though in practice this is unlikely to ever materially
    // block since processing results is CPU bound.
    let (tx_results, rx_results) = mpsc::unbounded_channel();

    let reporting_task = ReportingTask::new(rx_results, reporters);
    let task_handle = tokio::spawn(reporting_task.work_forever());

    let proxy = ReportingProxy { tx_results };
    let handle = ReportingTaskHandle {
        handle: task_handle,
    };

    (proxy, handle)
}
