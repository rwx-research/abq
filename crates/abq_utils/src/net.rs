//! Server/client interfaces for use among entities in the ABQ ecosystem.

pub(crate) mod tcp;
#[cfg(not(feature = "tls"))]
pub type ServerListener = tcp::ServerListener;
#[cfg(not(feature = "tls"))]
pub type ServerStream = tcp::Stream;
#[cfg(not(feature = "tls"))]
pub type ConfiguredClient = tcp::ConfiguredClient;
#[cfg(not(feature = "tls"))]
pub type ClientStream = tcp::Stream;

#[cfg(feature = "tls")]
pub(crate) mod tls;
#[cfg(feature = "tls")]
pub type ServerListener = tls::ServerListener;
#[cfg(feature = "tls")]
pub type ServerStream = tls::ServerStream;
#[cfg(feature = "tls")]
pub type ConfiguredClient = tls::ConfiguredClient;
#[cfg(feature = "tls")]
pub type ClientStream = tls::ClientStream;
