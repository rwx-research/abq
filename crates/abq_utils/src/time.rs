use serde_derive::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[repr(transparent)]
pub struct EpochMillis(u128);

impl EpochMillis {
    pub fn now() -> Self {
        Self(
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_millis(),
        )
    }

    pub fn as_millis(self) -> u128 {
        self.0
    }

    pub fn from_millis(millis: u128) -> Self {
        Self(millis)
    }
}
