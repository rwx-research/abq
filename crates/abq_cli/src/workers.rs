use std::num::{NonZeroU64, NonZeroUsize};
use std::path::PathBuf;
use std::time::Duration;

use abq_reporting::{CompletedSummary, Reporter};
use abq_utils::exit::ExitCode;
use abq_utils::net_protocol::entity::{RunnerMeta, WorkerTag};
use abq_utils::net_protocol::queue::InvokeWork;
use abq_utils::net_protocol::runners::CapturedOutput;
use abq_utils::net_protocol::workers::{RunId, RunnerKind};
use abq_workers::negotiate::{
    NegotiatedWorkers, QueueNegotiatorHandle, WorkersConfig, WorkersNegotiator,
};
use abq_workers::workers::{WorkerContext, WorkersExit, WorkersExitStatus};

mod reporting;
mod retry_manifest_tracker;
mod summary;

use futures::stream::StreamExt;
use signal_hook::consts::TERM_SIGNALS;
use signal_hook_tokio::Signals;

use crate::reporting::{ReporterKind, StdoutPreferences};
use crate::workers::reporting::create_reporting_task;

use self::reporting::{build_reporters, ReportingTaskHandle};

type ClientOptions = abq_utils::net_opt::ClientOptions<abq_utils::auth::User>;

pub fn start_workers_standalone(
    run_id: RunId,
    tag: WorkerTag,
    num_workers: NonZeroUsize,
    runner_kind: RunnerKind,
    working_dir: PathBuf,
    reporter_kinds: Vec<ReporterKind>,
    stdout_preferences: StdoutPreferences,
    batch_size: NonZeroU64,
    test_results_timeout: Duration,
    queue_negotiator: QueueNegotiatorHandle,
    client_opts: ClientOptions,
) -> ! {
    let rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    let test_suite_name = "suite"; // TODO: determine this correctly
    let reporters = build_reporters(reporter_kinds, stdout_preferences, test_suite_name);

    rt.block_on(start_standalone_help(
        run_id,
        tag,
        num_workers,
        runner_kind,
        working_dir,
        reporters,
        stdout_preferences,
        batch_size,
        test_results_timeout,
        queue_negotiator,
        client_opts,
    ))
}

async fn start_standalone_help(
    run_id: RunId,
    tag: WorkerTag,
    num_workers: NonZeroUsize,
    runner_kind: RunnerKind,
    working_dir: PathBuf,
    reporters: Vec<Box<dyn Reporter>>,
    stdout_preferences: StdoutPreferences,
    batch_size: NonZeroU64,
    test_results_timeout: Duration,
    queue_negotiator: QueueNegotiatorHandle,
    client_opts: ClientOptions,
) -> ! {
    let mut term_signals = Signals::new(TERM_SIGNALS).unwrap();

    let context = WorkerContext::AlwaysWorkIn { working_dir };

    let (reporting_proxy, reporting_handle) = create_reporting_task(reporters);

    let workers_config = WorkersConfig {
        tag,
        num_workers,
        runner_kind,
        local_results_handler: Box::new(reporting_proxy),
        worker_context: context,
        debug_native_runner: std::env::var_os("ABQ_DEBUG_NATIVE").is_some(),
        results_batch_size_hint: batch_size.get(),
    };

    tracing::debug!(
        "Workers attaching to queue negotiator {}",
        queue_negotiator.get_address()
    );

    let invoke_work = InvokeWork {
        run_id: run_id.clone(),
        batch_size_hint: batch_size,
        test_results_timeout,
    };

    let mut worker_pool = WorkersNegotiator::negotiate_and_start_pool_on_executor(
        workers_config,
        queue_negotiator,
        client_opts,
        invoke_work,
    )
    .await
    .unwrap();

    tracing::debug!("Workers attached");

    // Shut down the pool when
    //   - all its workers are done
    //   - we get a signal to shutdown
    loop {
        tokio::select! {
            () = worker_pool.wait() => {
                do_shutdown(worker_pool, reporting_handle, stdout_preferences).await;
            }
            _ = term_signals.next() => {
                do_cancellation_shutdown(worker_pool);
            }
        }
    }
}

async fn do_shutdown(
    mut worker_pool: NegotiatedWorkers,
    reporting_handle: ReportingTaskHandle,
    stdout_preferences: StdoutPreferences,
) -> ! {
    let WorkersExit {
        status,
        native_runner_info,
        manifest_generation_output,
        final_captured_outputs,
    } = worker_pool.shutdown();

    tracing::debug!("Workers shutdown");

    let finalized_reporters = reporting_handle.join().await;

    if let Some((runner, manifest_output)) = manifest_generation_output {
        print_manifest_generation_output(runner, manifest_output);
    }
    print_final_runner_outputs(final_captured_outputs);

    let completed_summary = CompletedSummary { native_runner_info };

    let (suite_result, errors) = finalized_reporters.finish(&completed_summary);

    for error in errors {
        eprintln!("{error}");
    }

    print!("\n\n");
    suite_result
        .write_short_summary_lines(&mut stdout_preferences.stdout_stream())
        .unwrap();

    // If the workers didn't fault, exit with whatever status the test suite run is at; otherwise,
    // indicate the worker fault.
    let exit_code = match status {
        WorkersExitStatus::Completed(runner_exit_code) => {
            suite_result.suggested_exit_code().max(runner_exit_code)
        }
        WorkersExitStatus::Error { errors } => {
            for error in errors {
                eprintln!("{error}");
            }
            ExitCode::ABQ_ERROR
        }
    };

    std::process::exit(exit_code.get());
}

fn do_cancellation_shutdown(worker_pool: NegotiatedWorkers) -> ! {
    let WorkersExit {
        status: _,
        native_runner_info: _,
        manifest_generation_output: _,
        final_captured_outputs,
    } = worker_pool.cancel();

    tracing::debug!("Workers cancelled");

    print_final_runner_outputs(final_captured_outputs);

    std::process::exit(ExitCode::CANCELLED.get());
}

pub(crate) fn print_manifest_generation_output(
    runner: RunnerMeta,
    manifest_output: CapturedOutput,
) {
    // NB: there is no reasonable way to surface the error at this point, since we are
    // about to shut down.
    let _opt_err = abq_reporting::output::format_manifest_generation_output(
        &mut std::io::stdout(),
        runner,
        &manifest_output,
    );
}

pub(crate) fn print_final_runner_outputs(
    final_captured_runner_outputs: Vec<(RunnerMeta, CapturedOutput)>,
) {
    for (runner, runner_out) in final_captured_runner_outputs {
        // NB: there is no reasonable way to surface the error at this point, since we are
        // about to shut down.
        let _opt_err = abq_reporting::output::format_final_runner_output(
            &mut std::io::stdout(),
            runner,
            &runner_out,
        );
    }
}
