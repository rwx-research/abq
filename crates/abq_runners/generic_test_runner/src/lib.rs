use std::io;
use std::net::TcpStream;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};
use std::{net::TcpListener, process};

use abq_utils::net_protocol::runners::{
    AbqProtocolVersionMessage, ManifestMessage, TestCaseMessage, TestResult, TestResultMessage,
    ABQ_GENERATE_MANIFEST, ABQ_SOCKET, ACTIVE_PROTOCOL_VERSION_MAJOR,
    ACTIVE_PROTOCOL_VERSION_MINOR,
};
use abq_utils::net_protocol::workers::{
    InvocationId, NativeTestRunnerParams, NextWork, WorkContext, WorkId,
};
use abq_utils::{flatten_manifest, net_protocol};
use thiserror::Error;
use tracing::instrument;

pub struct GenericTestRunner;

static POLL_WAIT_TIME: Duration = Duration::from_millis(10);

#[derive(Debug, Error)]
pub enum ProtocolVersionError {
    #[error("Timeout while waiting for protocol version")]
    Timeout,
    #[error("Incompatible native runner protocol version")]
    NotCompatible,
}

#[derive(Debug, Error)]
pub enum ProtocolVersionMessageError {
    #[error("{0}")]
    Io(#[from] io::Error),
    #[error("{0}")]
    Version(#[from] ProtocolVersionError),
}

type RunnerConnection = TcpStream;

/// Opens a connection with native test runner on the given listener.
///
/// Validates that the connected native runner communicates with a compatible ABQ protocol.
/// Returns an error if the protocols are incompatible, or the process fails to establish a
/// connection in a suitable time frame.
pub fn open_native_runner_connection(
    listener: &mut TcpListener,
    timeout: Duration,
) -> Result<RunnerConnection, ProtocolVersionMessageError> {
    use ProtocolVersionMessageError::*;

    listener.set_nonblocking(true).map_err(Io)?;

    let start = Instant::now();
    let (
        AbqProtocolVersionMessage {
            r#type: _,
            major,
            minor,
        },
        runner_conn,
    ) = loop {
        if start.elapsed() >= timeout {
            return Err(ProtocolVersionError::Timeout.into());
        }

        match listener.accept() {
            Ok((mut conn, _)) => {
                conn.set_nonblocking(false).map_err(Io)?;
                listener.set_nonblocking(false).map_err(Io)?;

                let version_message: AbqProtocolVersionMessage =
                    net_protocol::read(&mut conn).map_err(Io)?;

                break (version_message, conn);
            }
            Err(err) => match err.kind() {
                io::ErrorKind::WouldBlock => {
                    std::thread::sleep(POLL_WAIT_TIME);
                    continue;
                }
                _ => return Err(Io(err)),
            },
        }
    };

    if major != ACTIVE_PROTOCOL_VERSION_MAJOR || minor != ACTIVE_PROTOCOL_VERSION_MINOR {
        Err(ProtocolVersionError::NotCompatible.into())
    } else {
        Ok(runner_conn)
    }
}

pub fn wait_for_manifest(mut runner_conn: RunnerConnection) -> io::Result<ManifestMessage> {
    let manifest: ManifestMessage = net_protocol::read(&mut runner_conn)?;

    Ok(manifest)
}

#[derive(Debug, Error)]
enum RetrieveManifestError {
    #[error("{0}")]
    Io(#[from] io::Error),
    #[error("{0}")]
    ProtocolVersion(#[from] ProtocolVersionError),
}

impl From<ProtocolVersionMessageError> for RetrieveManifestError {
    fn from(e: ProtocolVersionMessageError) -> Self {
        match e {
            ProtocolVersionMessageError::Io(e) => Self::Io(e),
            ProtocolVersionMessageError::Version(e) => Self::ProtocolVersion(e),
        }
    }
}

/// Retrieves the test manifest from native test runner.
#[instrument(level = "trace", skip(additional_env, working_dir))]
fn retrieve_manifest<'a>(
    cmd: &str,
    args: &[String],
    additional_env: impl IntoIterator<Item = (&'a String, &'a String)>,
    working_dir: &Path,
    protocol_version_timeout: Duration,
) -> Result<ManifestMessage, RetrieveManifestError> {
    // One-shot the native runner. Since we set the manifest generation flag, expect exactly one
    // message to be received, namely the manifest.
    let manifest = {
        let mut our_listener = TcpListener::bind("127.0.0.1:0")?;
        let our_addr = our_listener.local_addr()?;

        let mut native_runner = process::Command::new(cmd);
        native_runner.args(args);
        native_runner.env(ABQ_SOCKET, format!("{}", our_addr));
        native_runner.env(ABQ_GENERATE_MANIFEST, "1");
        native_runner.envs(additional_env);
        native_runner.current_dir(working_dir);
        native_runner.stdout(process::Stdio::null());
        native_runner.stderr(process::Stdio::null());
        let mut native_runner_handle = native_runner.spawn()?;

        // Wait for and validate the protocol version message, which must always come first.
        let runner_conn =
            open_native_runner_connection(&mut our_listener, protocol_version_timeout)?;

        let manifest = wait_for_manifest(runner_conn)?;

        let status = native_runner_handle.wait()?;
        debug_assert!(status.success());

        manifest
    };

    Ok(manifest)
}

#[derive(Debug, Error)]
pub enum RunnerError {
    #[error("{0}")]
    Io(io::Error),
    #[error("{0}")]
    ProtocolVersion(ProtocolVersionError),
}

impl From<RetrieveManifestError> for RunnerError {
    fn from(e: RetrieveManifestError) -> Self {
        match e {
            RetrieveManifestError::Io(e) => Self::Io(e),
            RetrieveManifestError::ProtocolVersion(e) => Self::ProtocolVersion(e),
        }
    }
}

impl From<ProtocolVersionMessageError> for RunnerError {
    fn from(e: ProtocolVersionMessageError) -> Self {
        match e {
            ProtocolVersionMessageError::Io(e) => Self::Io(e),
            ProtocolVersionMessageError::Version(e) => Self::ProtocolVersion(e),
        }
    }
}

impl GenericTestRunner {
    pub fn run<ShouldShutdown, SendManifest, GetNextWork, SendTestResult>(
        input: NativeTestRunnerParams,
        working_dir: &Path,
        _polling_should_shutdown: ShouldShutdown,
        send_manifest: Option<SendManifest>,
        mut get_next_test: GetNextWork,
        mut send_test_result: SendTestResult,
    ) -> Result<(), RunnerError>
    where
        ShouldShutdown: Fn() -> bool,
        SendManifest: FnMut(ManifestMessage),
        GetNextWork: FnMut() -> NextWork + std::marker::Send + 'static,
        SendTestResult: FnMut(WorkId, TestResult),
    {
        let NativeTestRunnerParams {
            cmd,
            args,
            extra_env: additional_env,
        } = input;

        // TODO: get from runner params
        let protocol_version_timeout = Duration::from_secs(5);

        // If we need to retrieve the manifest, do that first.
        if let Some(mut send_manifest) = send_manifest {
            // TODO: error handling
            let manifest = retrieve_manifest(
                &cmd,
                &args,
                &additional_env,
                working_dir,
                protocol_version_timeout,
            )?;

            send_manifest(manifest);
        }

        // Now, start the test runner.
        //
        // The interface between us and the runner is described at
        //   https://www.notion.so/rwx/ABQ-Worker-Native-Test-Runner-Interface-0959f5a9144741d798ac122566a3d887
        //
        // In short: the protocol is
        //
        // Queue |                       | Worker (us) |                           | Native runner
        //       | <- send Next-Test -   |             |                           |
        //       | - recv Next-Test ->   |             |                           |
        //       |                       |             | - send TestCaseMessage -> |
        //       |                       |             | <- recv TestResult -      |
        //       | <- send Test-Result - |             |                           |
        //       |                       |             |                           |
        //       |          ...          |             |          ...              |
        //       |                       |             |                           |
        //       | <- send Next-Test -   |             |                           |
        //       | -   recv Done    ->   |             |      <close conn>         |
        // Queue |                       | Worker (us) |                           | Native runner
        let mut our_listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let our_addr = our_listener.local_addr().unwrap();

        let mut native_runner = process::Command::new(cmd);
        native_runner.args(args);
        native_runner.env(ABQ_SOCKET, format!("{}", our_addr));
        native_runner.envs(additional_env);
        native_runner.current_dir(working_dir);
        native_runner.stdout(process::Stdio::null());
        native_runner.stderr(process::Stdio::null());
        let mut native_runner_handle = native_runner.spawn().unwrap();

        // First, get and validate the protocol version message.
        let mut runner_conn =
            open_native_runner_connection(&mut our_listener, protocol_version_timeout)?;

        // We establish one connection with the native runner and repeatedly send tests until we're
        // done.
        loop {
            match get_next_test() {
                NextWork::EndOfWork => {
                    drop(runner_conn);
                    break;
                }
                NextWork::Work {
                    test_case,
                    context: _,
                    invocation_id: _,
                    work_id,
                } => {
                    let test_case_message = TestCaseMessage { test_case };

                    // TODO: errors
                    net_protocol::write(&mut runner_conn, test_case_message).unwrap();
                    let test_result_message: TestResultMessage =
                        net_protocol::read(&mut runner_conn).unwrap();

                    send_test_result(work_id, test_result_message.test_result);
                }
            };
        }

        drop(our_listener);
        native_runner_handle.wait().unwrap();

        Ok(())
    }
}

/// Executes a native test runner in an end-to-end fashion from the perspective of an ABQ worker.
/// Returns the manifest and all test results.
pub fn execute_wrapped_runner(
    native_runner_params: NativeTestRunnerParams,
    working_dir: PathBuf,
) -> Result<(ManifestMessage, Vec<TestResult>), RunnerError> {
    let mut test_case_index = 0;

    let mut manifest_message = None;

    // Currently, an atomic mutex is used here because `send_manifest`, `get_next_test`, and the
    // main thread all need access to manifest's memory location. We can get away with unsafe code
    // here because we know that the manifest must come in before `get_next_test` will be called,
    // and moreover, `send_manifest` will never be called again. But to avoid bugs, we don't do
    // that for now.
    let flat_manifest = Arc::new(Mutex::new(None));

    let mut test_results = vec![];

    let send_manifest = {
        let flat_manifest = Arc::clone(&flat_manifest);
        let manifest_message = &mut manifest_message;
        move |real_manifest: ManifestMessage| {
            let mut flat_manifest = flat_manifest.lock().unwrap();

            if manifest_message.is_some() || flat_manifest.is_some() {
                panic!("Manifest has already been defined, but is being sent again");
            }

            *manifest_message = Some(real_manifest.clone());
            *flat_manifest = Some(flatten_manifest(real_manifest.manifest));
        }
    };

    let get_next_test = {
        let manifest = Arc::clone(&flat_manifest);
        let working_dir = working_dir.clone();
        move || {
            loop {
                let manifest = manifest.lock().unwrap();
                let next_test = match &(*manifest) {
                    Some(manifest) => manifest.get(test_case_index),
                    None => {
                        // still waiting for the manifest, spin
                        continue;
                    }
                };
                return match next_test {
                    Some(test_case) => {
                        test_case_index += 1;

                        NextWork::Work {
                            test_case: test_case.clone(),
                            context: WorkContext {
                                working_dir: working_dir.clone(),
                            },
                            invocation_id: InvocationId::new(),
                            work_id: WorkId(Default::default()),
                        }
                    }
                    None => NextWork::EndOfWork,
                };
            }
        }
    };
    let send_test_result = |_, test_result| test_results.push(test_result);

    GenericTestRunner::run(
        native_runner_params,
        &working_dir,
        || false,
        Some(send_manifest),
        get_next_test,
        send_test_result,
    )?;

    Ok((
        manifest_message.expect("manifest never received!"),
        test_results,
    ))
}

#[cfg(test)]
mod test_validate_protocol_version_message {
    use std::{
        net::{TcpListener, TcpStream},
        thread,
        time::Duration,
    };

    use abq_utils::net_protocol::{
        self,
        runners::{
            AbqProtocolVersionMessage, AbqProtocolVersionTag, ACTIVE_PROTOCOL_VERSION_MAJOR,
            ACTIVE_PROTOCOL_VERSION_MINOR,
        },
    };

    use super::{open_native_runner_connection, ProtocolVersionError, ProtocolVersionMessageError};

    #[test]
    fn recv_and_validate_protocol_version_message() {
        let mut listener = TcpListener::bind("0.0.0.0:0").unwrap();
        let socket_addr = listener.local_addr().unwrap();
        thread::spawn(move || {
            let mut stream = TcpStream::connect(socket_addr).unwrap();
            let version_message = AbqProtocolVersionMessage {
                r#type: AbqProtocolVersionTag::AbqProtocolVersion,
                major: ACTIVE_PROTOCOL_VERSION_MAJOR,
                minor: ACTIVE_PROTOCOL_VERSION_MINOR,
            };
            net_protocol::write(&mut stream, version_message).unwrap();
        })
        .join()
        .unwrap();

        let result = open_native_runner_connection(&mut listener, Duration::from_secs(1));

        assert!(result.is_ok());
    }

    #[test]
    fn protocol_version_message_incompatible() {
        let mut listener = TcpListener::bind("0.0.0.0:0").unwrap();
        let socket_addr = listener.local_addr().unwrap();
        thread::spawn(move || {
            let mut stream = TcpStream::connect(socket_addr).unwrap();
            let version_message = AbqProtocolVersionMessage {
                r#type: AbqProtocolVersionTag::AbqProtocolVersion,
                major: 999123123,
                minor: 12312342,
            };
            net_protocol::write(&mut stream, version_message).unwrap();
        })
        .join()
        .unwrap();

        let result = open_native_runner_connection(&mut listener, Duration::from_secs(1));

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(
            err,
            ProtocolVersionMessageError::Version(ProtocolVersionError::NotCompatible)
        ));
    }

    #[test]
    fn protocol_version_message_recv_wrong_message() {
        let mut listener = TcpListener::bind("0.0.0.0:0").unwrap();
        let socket_addr = listener.local_addr().unwrap();
        thread::spawn(move || {
            let mut stream = TcpStream::connect(socket_addr).unwrap();
            let message = net_protocol::runners::Manifest { members: vec![] };
            net_protocol::write(&mut stream, message).unwrap();
        })
        .join()
        .unwrap();

        let result = open_native_runner_connection(&mut listener, Duration::from_secs(1));

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(err, ProtocolVersionMessageError::Io(_)));
    }

    #[test]
    fn protocol_version_message_tunnel_dropped() {
        let mut listener = TcpListener::bind("0.0.0.0:0").unwrap();
        let socket_addr = listener.local_addr().unwrap();
        thread::spawn(move || {
            let stream = TcpStream::connect(socket_addr).unwrap();
            drop(stream);
        })
        .join()
        .unwrap();

        let result = open_native_runner_connection(&mut listener, Duration::from_secs(1));

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(err, ProtocolVersionMessageError::Io(_)));
    }

    #[test]
    fn protocol_version_message_tunnel_timeout() {
        let mut listener = TcpListener::bind("0.0.0.0:0").unwrap();
        let socket_addr = listener.local_addr().unwrap();

        let timeout = Duration::from_millis(0);
        thread::spawn(move || {
            thread::sleep(Duration::from_millis(100));
            let mut stream = TcpStream::connect(socket_addr).unwrap();
            let version_message = AbqProtocolVersionMessage {
                r#type: AbqProtocolVersionTag::AbqProtocolVersion,
                major: ACTIVE_PROTOCOL_VERSION_MAJOR,
                minor: ACTIVE_PROTOCOL_VERSION_MINOR,
            };
            net_protocol::write(&mut stream, version_message).unwrap();
        })
        .join()
        .unwrap();

        let result = open_native_runner_connection(&mut listener, timeout);

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(
            err,
            ProtocolVersionMessageError::Version(ProtocolVersionError::Timeout)
        ));
    }
}

#[cfg(test)]
#[cfg(feature = "test-abq-jest")]
mod test_abq_jest {
    use crate::{execute_wrapped_runner, GenericTestRunner};
    use abq_utils::net_protocol::runners::{ManifestMessage, Status, TestOrGroup};
    use abq_utils::net_protocol::workers::{NativeTestRunnerParams, NextWork};

    use std::path::PathBuf;

    fn npm_jest_project_path() -> PathBuf {
        PathBuf::from(std::env::var("ABQ_WORKSPACE_DIR").unwrap())
            .join("testdata/jest/npm-jest-project")
    }

    #[test]
    fn get_manifest() {
        let input = NativeTestRunnerParams {
            cmd: "npm".to_string(),
            args: vec!["test".to_string()],
            extra_env: Default::default(),
        };

        let mut manifest = None;
        let mut test_results = vec![];

        let send_manifest = |real_manifest| manifest = Some(real_manifest);
        let get_next_test = || NextWork::EndOfWork;
        let send_test_result = |_, test_result| test_results.push(test_result);

        GenericTestRunner::run(
            input,
            &npm_jest_project_path(),
            || false,
            Some(send_manifest),
            get_next_test,
            send_test_result,
        )
        .unwrap();

        assert!(test_results.is_empty());

        let ManifestMessage { mut manifest } = manifest.unwrap();

        manifest.members.sort_by_key(|member| match member {
            TestOrGroup::Test(test) => test.id.clone(),
            TestOrGroup::Group(group) => group.name.clone(),
        });

        insta::assert_json_snapshot!(manifest);
    }

    #[test]
    fn get_manifest_and_run_tests() {
        let working_dir = npm_jest_project_path();
        let input = NativeTestRunnerParams {
            cmd: "npm".to_string(),
            args: vec!["test".to_string()],
            extra_env: Default::default(),
        };

        let (_, mut test_results) = execute_wrapped_runner(input, working_dir).unwrap();

        test_results.sort_by_key(|r| r.id.clone());

        assert_eq!(test_results.len(), 2, "{:#?}", test_results);

        assert_eq!(test_results[0].status, Status::Success);
        assert!(test_results[0].id.ends_with("add.test.js"));

        assert_eq!(test_results[1].status, Status::Success);
        assert!(test_results[1].id.ends_with("names.test.js"));
    }
}
