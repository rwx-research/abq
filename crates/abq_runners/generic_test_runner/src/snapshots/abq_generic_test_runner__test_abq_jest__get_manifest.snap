---
source: crates/abq_runners/generic_test_runner/src/lib.rs
expression: manifest
---
{
  "members": [
    {
      "type": "test",
      "id": "add.test.js",
      "tags": [],
      "meta": {
        "fileName": "add.test.js"
      }
    },
    {
      "type": "test",
      "id": "names.test.js",
      "tags": [],
      "meta": {
        "fileName": "names.test.js"
      }
    }
  ],
  "init_meta": {}
}
